crd/crd.md

```diff
@@ -0,0 +1,231 @@
+# Kubernetes Custom Resources
+
+Custom resources allow to extend Kubernetes API with additional types not
+available in the default Kubernetes distribution.
+
+Like native types, such as `Pod`, `Deployment` or `Service`, custom resources
+become first-class citizens in a Kubernetes cluster and can be queried via API
+and manipulated by `kubectl`.
+
+## Creating Custom Resource Definition
+
+Before a custom resource can be created, we need to let Kubernetes know what
+the new resource kind is, what its spec looks like, how to validate its fields
+and so on.
+
+Kubernetes provides an easy declarative way of doing so via a special resource
+called `CustomResourceDefinition`, or CRD. Let's create a CRD spec that
+describes a new resources called `Nginx`:
+
+```shell
+$ cat ~/workshop/crd/assets/crd.yaml
+```
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: nginxes.training.example.com
+spec:
+  group: training.example.com
+  version: v1
+  scope: Namespaced
+  names:
+    kind: Nginx
+    plural: nginxes
+    singular: nginx
+    shortNames:
+      - ng
+  validation:
+    openAPIV3Schema:
+      required: ["spec"]
+      properties:
+        spec:
+          required: ["version"]
+          properties:
+            version:
+              type: "string"
+```
+
+## Registering Custom Resource
+
+Now that the spec for our new resource has been defined, we need to register
+it with the Kubernetes API server.
+
+To do that, simply create the CRD resource:
+
+```shell
+$ kubectl create -f ~/workshop/crd/assets/crd.yaml
+```
+
+CRDs themselves are global (not namespaced) resources and all currently
+registered custom types can be viewed using:
+
+```shell
+$ kubectl get crd
+```
+
+## Defining Custom Resource
+
+Now that the custom type has been registered, we can create our custom resource.
+
+Let's define its spec:
+
+```shell
+$ cat ~/workshop/crd/assets/nginx.yaml
+```
+
+```yaml
+apiVersion: training.example.com/v1
+kind: Nginx
+metadata:
+  name: mynginx
+spec:
+  version: 1.9.1
+```
+
+## Creating Custom Resource
+
+As already mentioned, after having been registered, custom resources are
```

Comment:
Since custom resources are treated exactly as built-in resources, to create a new custom resource we can use: